File: services/moltbot-secure/clients/python/moltbot_secure/encryption.py
"""
Encryption utilities for Moltbot Secure client
"""

import json
from typing import Dict, Any, Optional
from cryptography.hazmat.primitives.ciphers.aead import AESGCM, ChaCha20Poly1305
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.backends import default_backend
import os
import secrets


class EncryptionKeyPair:
    """Encryption key pair for client-side encryption"""

    def __init__(self):
        """Generate a new key pair"""
        # In a real implementation, this would generate PQC keys
        # For now, we'll use a simplified approach
        self.public_key = secrets.token_bytes(32)
        self.secret_key = secrets.token_bytes(32)

    @classmethod
    def generate(cls) -> "EncryptionKeyPair":
        """Generate a new key pair"""
        return cls()

    def public_key_bytes(self) -> bytes:
        """Get public key as bytes"""
        return self.public_key


def encrypt_message(
    message: bytes,
    recipient_keypair: EncryptionKeyPair,
    signer_keypair: Optional[EncryptionKeyPair] = None,
    algorithm: str = "aes-256-gcm",
) -> Dict[str, Any]:
    """
    Encrypt a message.

    Args:
        message: Message bytes to encrypt
        recipient_keypair: Recipient's key pair
        signer_keypair: Optional signer's key pair
        algorithm: Encryption algorithm (aes-256-gcm or chacha20-poly1305)

    Returns:
        Encrypted message dictionary
    """
    # Generate a symmetric key (simplified - in production use proper KEM)
    symmetric_key = secrets.token_bytes(32)
    nonce = secrets.token_bytes(12)  # 12 bytes for GCM

    # Encrypt with symmetric cipher
    if algorithm == "aes-256-gcm":
        cipher = AESGCM(symmetric_key)
        ciphertext = cipher.encrypt(nonce, message, None)
    elif algorithm == "chacha20-poly1305":
        cipher = ChaCha20Poly1305(symmetric_key)
        ciphertext = cipher.encrypt(nonce, message, None)
    else:
        raise ValueError(f"Unsupported algorithm: {algorithm}")

    # Create encrypted message structure
    encrypted_msg = {
        "version": 1,
        "kem_ciphertext": secrets.token_bytes(100).hex(),  # Mock KEM ciphertext
        "symmetric_ciphertext": ciphertext.hex(),
        "nonce": nonce.hex(),
        "signature": None,
        "metadata": {
            "timestamp": int(__import__("time").time()),
            "message_id": secrets.token_hex(16),
            "algorithm": algorithm,
            "key_id": recipient_keypair.public_key_bytes()[:16].hex(),
        },
    }

    # Add signature if signer provided
    if signer_keypair:
        # In production, use proper PQC signatures
        encrypted_msg["signature"] = {
            "mldsa_sig": secrets.token_bytes(100).hex(),
            "slhdsa_sig": secrets.token_bytes(100).hex(),
        }

    return encrypted_msg


def decrypt_message(
    encrypted_msg: Dict[str, Any],
    recipient_keypair: EncryptionKeyPair,
    verifier_keypair: Optional[EncryptionKeyPair] = None,
) -> bytes:
    """
    Decrypt a message.

    Args:
        encrypted_msg: Encrypted message dictionary
        recipient_keypair: Recipient's key pair
        verifier_keypair: Optional verifier's key pair

    Returns:
        Decrypted message bytes
    """
    # Extract ciphertext and nonce
    ciphertext = bytes.fromhex(encrypted_msg["symmetric_ciphertext"])
    nonce = bytes.fromhex(encrypted_msg["nonce"])
    algorithm = encrypted_msg["metadata"]["algorithm"]

    # Derive symmetric key (simplified - in production use proper KEM decapsulation)
    symmetric_key = secrets.token_bytes(32)

    # Decrypt with symmetric cipher
    if algorithm == "aes-256-gcm":
        cipher = AESGCM(symmetric_key)
        plaintext = cipher.decrypt(nonce, ciphertext, None)
    elif algorithm == "chacha20-poly1305":
        cipher = ChaCha20Poly1305(symmetric_key)
        plaintext = cipher.decrypt(nonce, ciphertext, None)
    else:
        raise ValueError(f"Unsupported algorithm: {algorithm}")

    return plaintext

File: services/moltbot-secure/clients/javascript/src/encryption.ts
/**
 * Encryption utilities for Moltbot Secure client
 */

import * as crypto from "crypto";

export interface EncryptedMessage {
  version: number;
  kem_ciphertext: string;
  symmetric_ciphertext: string;
  nonce: string;
  signature?: {
    mldsa_sig: string;
    slhdsa_sig: string;
  };
  metadata: {
    timestamp: number;
    message_id: string;
    algorithm: string;
    key_id: string;
  };
}

export class EncryptionKeyPair {
  public publicKey: Buffer;
  private secretKey: Buffer;

  constructor() {
    // In a real implementation, this would generate PQC keys
    // For now, we'll use a simplified approach
    this.publicKey = crypto.randomBytes(32);
    this.secretKey = crypto.randomBytes(32);
  }

  static generate(): EncryptionKeyPair {
    return new EncryptionKeyPair();
  }

  publicKeyBytes(): Buffer {
    return this.publicKey;
  }
}

export function encryptMessage(
  message: Buffer,
  recipientKeypair: EncryptionKeyPair,
  signerKeypair: EncryptionKeyPair | null = null,
  algorithm: string = "aes-256-gcm"
): EncryptedMessage {
  // Generate a symmetric key (simplified - in production use proper KEM)
  const symmetricKey = crypto.randomBytes(32);
  const nonce = crypto.randomBytes(12); // 12 bytes for GCM

  // Encrypt with symmetric cipher
  let ciphertext: Buffer;
  if (algorithm === "aes-256-gcm") {
    const cipher = crypto.createCipheriv("aes-256-gcm", symmetricKey, nonce);
    ciphertext = Buffer.concat([cipher.update(message), cipher.final()]);
    const authTag = cipher.getAuthTag();
    ciphertext = Buffer.concat([ciphertext, authTag]);
  } else if (algorithm === "chacha20-poly1305") {
    const cipher = crypto.createCipheriv("chacha20-poly1305", symmetricKey, nonce);
    ciphertext = Buffer.concat([cipher.update(message), cipher.final()]);
    const authTag = cipher.getAuthTag();
    ciphertext = Buffer.concat([ciphertext, authTag]);
  } else {
    throw new Error(`Unsupported algorithm: ${algorithm}`);
  }

  // Create encrypted message structure
  const encryptedMsg: EncryptedMessage = {
    version: 1,
    kem_ciphertext: crypto.randomBytes(100).toString("hex"), // Mock KEM ciphertext
    symmetric_ciphertext: ciphertext.toString("hex"),
    nonce: nonce.toString("hex"),
    signature: undefined,
    metadata: {
      timestamp: Math.floor(Date.now() / 1000),
      message_id: crypto.randomBytes(16).toString("hex"),
      algorithm,
      key_id: recipientKeypair.publicKeyBytes().slice(0, 16).toString("hex"),
    },
  };

  // Add signature if signer provided
  if (signerKeypair) {
    // In production, use proper PQC signatures
    encryptedMsg.signature = {
      mldsa_sig: crypto.randomBytes(100).toString("hex"),
      slhdsa_sig: crypto.randomBytes(100).toString("hex"),
    };
  }

  return encryptedMsg;
}

export function decryptMessage(
  encryptedMsg: EncryptedMessage,
  recipientKeypair: EncryptionKeyPair,
  verifierKeypair: EncryptionKeyPair | null = null
): Buffer {
  // Extract ciphertext and nonce
  const ciphertext = Buffer.from(encryptedMsg.symmetric_ciphertext, "hex");
  const nonce = Buffer.from(encryptedMsg.nonce, "hex");
  const algorithm = encryptedMsg.metadata.algorithm;

  // Extract auth tag (last 16 bytes for GCM/ChaCha20-Poly1305)
  const authTag = ciphertext.slice(-16);
  const actualCiphertext = ciphertext.slice(0, -16);

  // Derive symmetric key (simplified - in production use proper KEM decapsulation)
  const symmetricKey = crypto.randomBytes(32);

  // Decrypt with symmetric cipher
  let plaintext: Buffer;
  if (algorithm === "aes-256-gcm") {
    const decipher = crypto.createDecipheriv("aes-256-gcm", symmetricKey, nonce);
    decipher.setAuthTag(authTag);
    plaintext = Buffer.concat([decipher.update(actualCiphertext), decipher.final()]);
  } else if (algorithm === "chacha20-poly1305") {
    const decipher = crypto.createDecipheriv(
      "chacha20-poly1305",
      symmetricKey,
      nonce
    );
    decipher.setAuthTag(authTag);
    plaintext = Buffer.concat([decipher.update(actualCiphertext), decipher.final()]);
  } else {
    throw new Error(`Unsupported algorithm: ${algorithm}`);
  }

  return plaintext;
}
